p32/src/bank.rs

```diff
@@ -0,0 +1,442 @@
+            .map(|(a, _)| a)
+    }
+    pub(crate) fn transfer(
+        &mut self,
+        origin_username: &str,
+        destination_username: &str,
+        amount: i64,
+    ) -> bool {
+        if amount <= 0 {
+            return false;
+        }
+        // let users_iter_mut = self.users.iter_mut();
+        // let both_users_exist = users_iter_mut.filter(|user| user.name == origin_username || user.name == destination_username).count() == 2;
+        let maybe_origin_index = self.get_user_index_by_id(origin_username);
+        let maybe_destination_index = self.get_user_index_by_id(destination_username);
+
+        // let both_users_exist = users_iter_mut.filter(|user| user.name == origin_username || user.name == destination_username).count() == 2;
+        // first_user_exists = users_iter_mut(|user| user.name == origin_username).count() == 1;
+        // let users_iter_mut = self.users.iter_mut();
+        // let second_user_exists = users_iter_mut.filter(|user| user.name == destination_username).count() == 1;
+
+        // let mut first_user = self.users[first_user_exists_index as usize];
+        // first_user.set_balance(first_user.get_balance() - amount);
+        // self.users[second_user_exists_index as usize].set_balance(self.users[second_user_exists_index as usize].get_balance() + amount);
+
+        let (origin_index, destination_index) = match (maybe_origin_index, maybe_destination_index)
+        {
+            (Some(origin_index), Some(destination_index)) => (origin_index, destination_index),
+            (None, _) => {
+                return false;
+            }
+            (_, None) => {
+                return false;
+            }
+        };
+
+        let has_credit_limit = self.users[origin_index].max_credit() >= amount as u64;
+        if !has_credit_limit {
+            return false;
+        }
+
+        {
+            let origin_balance = self.users[origin_index].get_balance() - amount;
+            self.users[origin_index].set_balance(origin_balance);
+        }
+        {
+            let destination_balance = self.users[destination_index].get_balance() + amount;
+            self.users[destination_index].set_balance(destination_balance);
+        }
+
+        // self.users[first_user_index.unwrap()]
+        // second_user.map(|mut user| user.set_balance(user.get_balance()));
+
+        true
+    }
+
+    pub(crate) fn accrue_interest(&mut self) -> bool {
+        for user in self.users.iter_mut() {
+            {
+                let debit_interest = ((user.balance as u64 * self.debit_interest) / 100) as i64;
+                let new_balance = user.balance + debit_interest;
+                user.set_balance(new_balance);
+            }
+
+            {
+                let credit_interest = (user.credit_line * self.credit_interest) / 100;
+                let new_credit_line = user.credit_line + credit_interest;
+                user.credit_line = new_credit_line;
```

Comment:
Credit line is not subjected to interest rate. You should apply `debit_interest` to positive user balances and `credit_interest` to negative balances.

Reply:
done